torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2103,24 +2109,19 @@ def state_dict(self, *args, **kwargs):
             else:
                 return {}
 
-        elif self._state_dict_type == StateDictType.LOCAL_STATE_DICT:
+        elif (
+            self._state_dict_type == StateDictType.LOCAL_STATE_DICT or
+            self._state_dict_type == StateDictType.SHARDED_STATE_DICT
```

Comment:
so it seems that sharded state dict calling `state_dict` is not meant to do much, we remove the checkpointed FLAT_PARAM and add new data to the state_dict for the sharded original parameters. 

If this is the case, could we just remove the super().state_dict calls, recurse ourselves and call the post hook?

Reply:
We need the recursive calls for 1.) constructing the correct prefix, 2.) calling post_hook in a reversed order. We can do this ourselves but it will be better to reuse state_dict logic. The only thing that sharded_state_dict does not need is the detach one but that should not causing too many overheads.